replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/modern/ReplicationPersistentStoreImpl.java

```diff
@@ -0,0 +1,160 @@
+    try {
+      Class<T> impl = interfaceMappings.get(klass);
+      return ConstructorUtils.invokeConstructor(impl);
+    } catch (Exception e) {
+      throw new ReplicationPersistenceException(
+          "Couldn't create new persistable " + klass.getName(), e);
+    }
+  }
+
+  @Override
+  public <T extends Persistable> Optional<T> get(Class<T> klass, String id)
+      throws ReplicationPersistenceException {
+    String cql = String.format("'id' = '%s'", id);
+    List<Map<String, Object>> storedMaps;
+    String typeName = getPersistenceType(klass);
+
+    try {
+      storedMaps = persistentStore.get(typeName, cql, DEFAULT_START_INDEX, DEFAULT_PAGE_SIZE);
+    } catch (PersistenceException e) {
+      throw new ReplicationPersistenceException(
+          String.format(
+              "An error occurred while trying to retrieve an object from persistence with class %s and id %s",
+              klass.getName(), id),
+          e);
+    }
+
+    if (storedMaps.isEmpty()) {
+      LOGGER.debug("Couldn't find persisted object of type {} with ID {}", typeName, id);
+      return Optional.empty();
+    } else {
+      if (storedMaps.size() > 1) {
+        LOGGER.debug("Found two objects in persistence of type {} with ID {}", typeName, id);
+      }
+
+      T persistable = createPersistable(klass);
+      persistable.readFromMap(storedMaps.get(0));
+      logIfInvalid(persistable);
+      return Optional.of(persistable);
+    }
+  }
+
+  @Override
+  public <T extends Persistable> Stream<T> get(Class<T> klass)
+      throws ReplicationPersistenceException {
+    List<Map<String, Object>> storedMaps;
+    String typeName = getPersistenceType(klass);
+    try {
+      storedMaps = persistentStore.get(typeName, "", DEFAULT_START_INDEX, DEFAULT_PAGE_SIZE);
+    } catch (PersistenceException e) {
+      throw new ReplicationPersistenceException(
+          "An error occurred while retrieving data of type " + typeName, e);
+    }
+
+    List<T> results = new ArrayList<>();
+    for (Map<String, Object> result : storedMaps) {
+      T persistable = createPersistable(klass);
+      persistable.readFromMap(result);
+      logIfInvalid(persistable);
+      results.add(persistable);
+    }
+
+    return results.stream();
+  }
+
+  private void logIfInvalid(Persistable persistable) {
+    if (!persistable.isValid()) {
+      LOGGER.debug(
+          "An invalid Persistable was loaded from the PersistentStore! id: {}",
+          persistable.getId());
+    }
+  }
+
+  @VisibleForTesting
+  String getPersistenceType(Class klass) throws ReplicationPersistenceException {
+    try {
+      return (String) FieldUtils.readStaticField(klass, "PERSISTENCE_TYPE");
+    } catch (Exception e) {
+      throw new ReplicationPersistenceException(
+          "Failed to retrieve the persistence type of the class " + klass.getName(), e);
+    }
+  }
+
+  @Override
+  public void save(Persistable object) throws ReplicationPersistenceException {
```

Comment:
An Idea that makes a lot of sense to me is that we should do all of our validation of data as it comes into our system. So I have validation in the _get_ methods. Should I be performing validation here as well? Obviously I don't want to save invalid data, but as long as we have proper validation elsewhere we shouldn't have to worry about it right?

Reply:
There are typically 2 levels of validation you see:
1- per attribute validation (e.g. timeout must be positive)
2- per object validation (e.g. if a given attribute is true then another one must be provided with something); in other words attribute dependencies between each other which from the valid state of an object.

I like the idea of validation as the data comes into the system. The question would be can we avoid doing that each time someone calls a getter?